packages/flutter_migrate/lib/src/commands/start.dart

```diff
@@ -0,0 +1,340 @@
+      List<Directory> directories = const <Directory>[]}) {
+    for (final Directory d in directories) {
+      try {
+        d.deleteSync(recursive: true);
+      } on FileSystemException catch (e) {
+        logger.printError(
+            'Unabled to delete ${d.path} due to ${e.message}, please clean up manually.');
+      }
+    }
+    for (final String p in paths) {
+      try {
+        fileSystem.directory(p).deleteSync(recursive: true);
+      } on FileSystemException catch (e) {
+        logger.printError(
+            'Unabled to delete $p due to ${e.message}, please clean up manually.');
+      }
+    }
+  }
+
+  bool _validateEnvironment(FlutterToolsEnvironment environment) {
+    if (environment.getString('FlutterProject.directory') == null) {
+      logger.printError(
+          'No valid flutter project found. This command must be run from a flutter project directory');
+      return false;
+    }
+    if (environment.getString('FlutterProject.manifest.appname') == null) {
+      logger.printError('No app name found in project pubspec.yaml');
+      return false;
+    }
+    if (!(environment.getBool('FlutterProject.android.exists') ?? false) &&
+        environment['FlutterProject.android.isKotlin'] == null) {
+      logger.printError(
+          'Could not detect if android project uses kotlin or java');
+      return false;
+    }
+    if (!(environment.getBool('FlutterProject.ios.exists') ?? false) &&
+        environment['FlutterProject.ios.isSwift'] == null) {
+      logger.printError(
+          'Could not detect if iosProject uses swift or objective-c');
+      return false;
+    }
+    return true;
+  }
+
+  /// Writes the files into the working directory for the developer to review and resolve any conflicts.
+  Future<void> writeStagingDir(MigrateResult migrateResult, Logger logger,
+      {bool verbose = false, required Directory projectRootDir}) async {
+    final Directory stagingDir =
+        projectRootDir.childDirectory(kDefaultMigrateStagingDirectoryName);
+    if (verbose) {
+      logger.printStatus(
+          'Writing migrate staging directory at `${stagingDir.path}`');
+    }
+    // Write files in working dir
+    for (final MergeResult result in migrateResult.mergeResults) {
+      final File file = stagingDir.childFile(result.localPath);
+      file.createSync(recursive: true);
+      if (result is StringMergeResult) {
+        file.writeAsStringSync(result.mergedString, flush: true);
+      } else {
+        file.writeAsBytesSync((result as BinaryMergeResult).mergedBytes,
+            flush: true);
+      }
+    }
+
+    // Write all files that are newly added in target
+    for (final FilePendingMigration addedFile in migrateResult.addedFiles) {
+      final File file = stagingDir.childFile(addedFile.localPath);
+      file.createSync(recursive: true);
+      try {
+        file.writeAsStringSync(addedFile.file.readAsStringSync(), flush: true);
+      } on FileSystemException {
```

Comment:
what type of FileSystemException would we hit? with reading or writing?

Reply:
This is to distinguish if we need to write/read as string or bytes. I couldnt find a nicer solution than this, but if you try to read/write bytedata as string, it throws, likewise with read/write string as bytes. This catches these cases and selects the alternative one.

Reply:
I bet we could do better if this didn't have to run on windows, but this is fine.